sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/index/Index.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.catalog.index;
+
+import java.util.Collections;
+import java.util.Map;
+
+import org.apache.spark.annotation.Evolving;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.expressions.FieldReference;
+
+/**
+ * An Index in a table
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public class Index {
+  private String indexName;
+  private String indexType;
+  private Identifier table;
+  private FieldReference[] columns;
+  private Map<String, String> properties = Collections.emptyMap();
+
+  public Index(
+      String indexName,
+      String indexType,
```

Comment:
What `indexType` could be? It is up to data source implementation?

Reply:
It's up to the data source implementation.

Reply:
Do we have example index type in mind? `String` seems vague to me as it can contain arbitrary stuff. Why don't we go with enum or class? Another perspective is why we cannot put the index type inside `properties`?

Reply:
The example index types are  `BLOOM_FILTER_INDEX`, `Z_ORDERING_INDEX`, `BTREE_INDEX`. I actually started with enum and changed to String. 
It's more convenient to have index type outside of `properties`. in majority of the data sources,  the create index syntax is 
`CREATE [index_type] INDEX index_name ON [TABLE] table_name (column_name [ , . . . ])[OPTIONS indexPropertyList]`

@cloud-fan Shall we use enum instead of String for index type? Even though the catalog implementation is responsible for recognizing and taking care of user-specified index type, we still need to add these specific index types in `SqlBase`? We probably want to use enum to let users know what index types are supported?

Reply:
Index is for performance only and Spark doesn't need to define the semantic, I think String is more convenient. Data source can define whatever index type they like and ask end-users to use. Spark doesn't need to be in the middle.